std/experimental/perpetual.d

```diff
@@ -0,0 +1,653 @@
+		 *  Forwards calls to opSlice(), length() etc
+		 */
+		@property Element[] Ref() { return _value; }
+
+		/// Convert to string the value, not the class itself
+		@property toString() { return to!string(_value); }
+
+		/// Index of uninitialized part of the array
+		@property initIndex() const { return _heap.unplowed/Element.sizeof; }
+	}
+	// scalar value or static array
+	else
+	{
+		alias Element=T;
+		private Element* _value;
+		private enum bool dynamic=false;
+		static assert(!hasIndirections!T
+		    , T.stringof~" is reference type");
+
+		/** Reference to underlying scalar value
+		 */
+		@property ref Element Ref() {	return *_value; }
+
+		/// Convert to string
+		@property toString() { return to!string(*_value); }
+	}
+
+	/** If whole or part of the file was created and 
+	 *  was initialized in constructor, return true. 
+	 *	For scalar types this means, the entire object was constructed.
+	 *  For dynamic arrays initIndex() shows first initialized
+	 *  elsement.
+	 */
+	@property master() const { return _heap.master; }
+
+	static if(isAssignable!Element)
+		enum mode=ShMem.Mode.readWrite;
+	else
+		enum mode=ShMem.Mode.readOnly;
+
+
+
+/**
+ * Universal constructor, for both scalar and array types.
+ * Opens file and assosiates object with it.
+ * The file is extended if required.
+ * The object is initialized if file was created or extended.
+ */
+	this(Arg...)(string path, Arg arg)
+	{
+		// dynamic array
+		static if(dynamic)
+		{
+			_heap=shMem(path, mode);
+			enforce(_heap.length >= Element.sizeof, _tag~": file is too small");
+			_value=cast(Element[]) _heap[0.._heap.length];
+			// initialization
+			static if(isAssignable!Element)
+			{
+				if(master)
+					foreach(i, ref x; _value[initIndex..$])
+						emplace(&x, arg);
+			}
+			else
+			{
+				static assert(arg.length == 0
+					, _tag~": attempt to initialize immutable memory");
+			}
+		}
+		// scalar value
+		else
+		{
+			_heap=shMem(path, Element.sizeof, mode);
+			enforce(_heap.length >= Element.sizeof, _tag~": file is too small");
+			_value=cast(Element*) _heap[].ptr;
+
+			// initialization
+			static if(isAssignable!Element)
+			{
+				if(master)
+					emplace(_value, arg);
+			}
+			else
+			{
+				static assert(arg.length == 0
+					, _tag~": attempt to initialize immutable object");
+			}
+		}
+	}
+
+/**
+ * Open file and map dynamic array to it.
+ * Creates array of requested length, file is extended if necessary
+ */
+	this(Arg...)(size_t len, string path, Arg arg)
+	if(dynamic)
+	{
+		immutable size_t size=len*Element.sizeof;
+		
+		_heap=shMem(path, size, mode);
+		_value=cast(Element[]) _heap[0..size];
+		// initialization
+		static if(isAssignable!Element)
+		{
+			if(master)
+				foreach(ref x; _value[initIndex..$])
+					emplace(&x, arg);
+		}
+		else
+		{
+			static assert(arg.length == 0
+				, _tag~": attempt to initialize immutable array");
+		}
+	}
+ 	
+
+ // get reference to wrapped object.
+	alias Ref this;
+
+}
+
+
+///
+unittest {
```

Comment:
This is a really long ddoc. Please consider breaking this into multiple unit tests. 

Also, please put the `{` on its own line.